datafusion/functions-aggregate/src/count.rs

```diff
@@ -433,6 +433,49 @@ impl GroupsAccumulator for CountGroupsAccumulator {
         Ok(vec![Arc::new(counts) as ArrayRef])
     }
 
+    fn convert_to_state(
+        &self,
+        values: &[ArrayRef],
+        opt_filter: Option<&BooleanArray>,
+    ) -> Result<Vec<ArrayRef>> {
+        let values = &values[0];
+
+        let state_array = match (values.logical_nulls(), opt_filter) {
+            (Some(nulls), None) => {
+                let mut builder = Int64Builder::with_capacity(values.len());
+                nulls
+                    .into_iter()
+                    .for_each(|is_valid| builder.append_value(is_valid as i64));
+                builder.finish()
+            }
+            (Some(nulls), Some(filter)) => {
+                let mut builder = Int64Builder::with_capacity(values.len());
+                nulls.into_iter().zip(filter.iter()).for_each(
+                    |(is_valid, filter_value)| {
+                        builder.append_value(
```

Comment:
`bitwise_and` + cast?

Reply:
maybe we can use the `nullif` kernel here

Something like
```rust
let nulls = and(nulls, not(filter));
let output = nullif(values);
```

Reply:
I came up wtih this in https://github.com/apache/datafusion/pull/11734:

```rust

/// Converts a `BooleanBuffer` representing a filter to a `NullBuffer`
/// where the NullBuffer is true for all values that were true
/// in the filter and `null` for any values that were false or null
fn filter_to_nulls(filter: &BooleanArray) -> Option<NullBuffer> {
    let (filter_bools, filter_nulls) = filter.clone().into_parts();
    // Only keep values where the filter was true
    // convert all false to null
    let filter_bools = NullBuffer::from(filter_bools);
    NullBuffer::union(Some(&filter_bools), filter_nulls.as_ref())
}

/// Compute the final null mask for an array
///
/// The output null mask :
/// * is true (non null) for all values that were true in the filter and non null in the input
/// * is false (null) for all values that were false in the filter or null in the input
fn filtered_null_mask(
    opt_filter: Option<&BooleanArray>,
    input: &dyn Array,
) -> Option<NullBuffer> {
    let opt_filter = opt_filter.and_then(filter_to_nulls);
    NullBuffer::union(opt_filter.as_ref(), input.nulls())
}
```

And then you compute the final null mask without messing with the input:

```rust

        let nulls = filtered_null_mask(opt_filter, sums);
        let sums = PrimitiveArray::<T>::new(sums.values().clone(), nulls)
            .with_data_type(self.sum_data_type.clone());
```

Reply:
Nice, using `NullBuffer::union` is much better for the readability

Reply:
That was the missing link for me (thank you!) -- we can operate directly on underlying buffers.

I've rewritten state conversion for count on bitand on buffers + cast to Int64 in the end, and according to benchmarks from the commit it got 20-25% faster.

Just a suggestion -- won't it be better to use BooleanBuffer + & (bitand operator) instead of NullBuffer + union? NullBuffer is a bit confusing, so I've "pulled" the logic from union right into state conversion function.

Additionally, I plan to prepare benches and minimize ArrayBuilder usage for min / max / sum during tomorrow.

Reply:
> I've rewritten state conversion for count on bitand on buffers + cast to Int64 in the end, and according to benchmarks from the commit it got 20-25% faster.

🎉 

> Just a suggestion -- won't it be better to use BooleanBuffer + & (bitand operator) instead of NullBuffer + union? NullBuffer is a bit confusing, so I've "pulled" the logic from union right into state conversion function.

I think they are equivalent:  [`NullBuffer`](https://docs.rs/arrow/latest/arrow/buffer/struct.NullBuffer.html) just wraps `BooleanBuffer` and `NullBuffer::union` just calls `&` underneath: https://docs.rs/arrow-buffer/52.2.0/src/arrow_buffer/buffer/null.rs.html#76 (after replicating the `match(nulls, filter)` logic)

I don't have a strong opinion about which is more/less confusing

What I suggest we do is pull the logic to compute the output null mask basd on the optional input nullmask and the optional filter into a function (like `fn filtered_null_mask`) as it will be used in basically all of the `convert_to_state` implementations. As long as it is well documented, I think either implementation will work well

> Additionally, I plan to prepare benches and minimize ArrayBuilder usage for min / max / sum during tomorrow.

Sounds good -- would you like to keep updating this PR or shall we merge this PR and continue improvements with additional PRs on main?

Reply:
I'd like to make these few changes in this PR (along with merging docs update and review suggestions) -- don't think it'll take long enough to accumulate any significant conflicts.

Reply:
Sounds good. We will wait for you to let us know.